virttest/qemu_devices/qdevices.py

```diff
@@ -1636,7 +1638,110 @@ def disable(self, monitor):
         return out, ver_out
 
 
-class QVirtioFSDev(QBaseDevice):
+class QDaemonDev(QBaseDevice):
+    """
+    Virtual daemon device.
+    """
+
+    def __init__(self, name, aobject, child_bus=None):
+        """
+        :param name: The daemon name.
+        :type name: str
+        :param aobject: The aobject of daemon.
+        :type aobject: str
+        :param child_bus: List of buses, which this device provides.
+        :type child_bus: QSparseBus
+        """
+        aid = '%s_%s' % (name, aobject)
+        super(QDaemonDev, self).__init__(name, aobject=aobject, child_bus=child_bus)
+        self.set_aid(aid)
+        self.set_param('name', name)
+        self._daemon_process = None
+
+    def is_daemon_alive(self):
+        """Check whether daemon is alive."""
+        return bool(self._daemon_process and self._daemon_process.is_alive())
+
+    def start_daemon(self):
+        """Start daemon."""
+        cmd = self.get_param('cmd')
+        name = self.get_param('name')
+        run_bg_kwargs = self.get_param('run_bg_kwargs', {})
+        status_active = self.get_param('status_active')
+        read_until_timeout = self.get_param('read_until_timeout', 5)
+        start_until_timeout = self.get_param('start_until_timeout', 1)
+
+        if cmd is None:
+            logging.warn('No provided command to start %s daemon.', name)
+            self._daemon_process = None
+
+        if self.is_daemon_alive():
+            return
+
+        logging.info('Running %s daemon command %s.', name, cmd)
+        self._daemon_process = aexpect.run_bg(cmd, **run_bg_kwargs)
+        if status_active:
+            self._daemon_process.read_until_any_line_matches(
+                    status_active, timeout=read_until_timeout)
+        else:
+            time.sleep(start_until_timeout)
```

Comment:
I suppose here `read_up_to_prompt` would be more proper.

Reply:
If so, how about introduce a function instead of status_active:
```
class QDaemonDev(QBaseDevice):
...
    def check_active():
        self._daemon_process.read_up_to_prompt()

class QVirtioFSDev(QBaseDevice):

    def check_active():
        self._daemon_process.read_until_any_line_matches(status_active,..)
```
And other child class(if there is some one) might implement their own one.

Or, at least for `QswtpmDev` it could have a `status_active = r"[\#\$]\s*$"`, instead of simple sleep.

Reply:
> Yes, But I am afraid that the swtpm daemon does not start at once even if catch the `r"[\#\$]\s*$"` when the host environment is a low resource.

In this case, sleep 1 can not assure it either, even worse that catch `r"[\#\$]\s*$"`. At least you should get the command return, before sleep.

Reply:
Hi @vivianQizhu 
All right, follow up you.
Thanks.

Reply:
> > If so, how about introduce a function instead of status_active:
> > ```
> > class QDaemonDev(QBaseDevice):
> > ...
> >     def check_active():
> >         self._daemon_process.read_up_to_prompt()

Hi @vivianQizhu 
the `read_up_to_prompt` is belongs to `class ShellSession(Expect):` which the` prompt=r"[\#\$]\s*$"` in the `def __init__(...)` , I means it works on the shell session scenario.
so I used the `read_until_last_line_matches` to replace the `read_up_to_prompt`, like:

```
    def check_active(self):
        """Check the active status of daemon."""
        status_active = self.get_param('status_active')
        read_until_timeout = self.get_param('read_until_timeout', 5)
        self._daemon_process.read_until_last_line_matches(status_active,
                                                          read_until_timeout)
```
Then follow your comment to catch the `r"[\#\$]\s*$"` in swtpm daemon, like:
```
class QswtpmDev(QDaemonDev):
...
        def start_daemon(self):
            self.set_param('status_active', r"[\#\$]\s*$")
```
But raised error:
```
  File "/home/yhong/for-support-tpm/kar/workspace/avocado-vt/virttest/qemu_devices/qdevices.py", line 1851, in start_daemon
    super(QswtpmDev, self).start_daemon()
  File "/home/yhong/for-support-tpm/kar/workspace/avocado-vt/virttest/qemu_devices/qdevices.py", line 1687, in start_daemon
    self.check_active()
  File "/home/yhong/for-support-tpm/kar/workspace/avocado-vt/virttest/qemu_devices/qdevices.py", line 1670, in check_active
    read_until_timeout)
  File "/home/yhong/for-support-tpm/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 931, in read_until_last_line_matches
    print_func)
  File "/home/yhong/for-support-tpm/kar/workspace/lib/python3.6/site-packages/aexpect-1.5.1-py3.6.egg/aexpect/client.py", line 867, in read_until_output_matches
    raise ExpectProcessTerminatedError(patterns, self.get_status(), o)
aexpect.exceptions.ExpectProcessTerminatedError: Process terminated while looking for patterns '[\\#\\$]\\s*$'    (status: 0,    output: '')
```
From the debug log, the output is empty and status is '0' if run a command successfully on the host, so I think we could not catch the `r"[\#\$]\s*$"`, but we could catch it under the shell session.
Maybe I misunderstand your comment, please point me, thanks.


> > 
> > class QVirtioFSDev(QBaseDevice):
> > 
> >     def check_active():
> >         self._daemon_process.read_until_any_line_matches(status_active,..)
> > ```
> 
> Hi @vivianQizhu
> I think that `read_until_any_line_matches` also could do the `read_up_to_prompt` job and define `read_up_to_prompt` to catch the last line in the base class maybe not common, I prefer the `read_until_any_line_matches` to check the status of daemon since it can cover the most situation, includes the `read_up_to_prompt`. however, will change it by your comments if you wanna.
> 
> > And other child class(if there is some one) might implement their own one.
> > Or, at least for `QswtpmDev` it could have a `status_active = r"[\#\$]\s*$"`, instead of simple sleep.
> 
> Yes, But I am afraid that the swtpm daemon does not start at once even if catch the `r"[\#\$]\s*$"` when the host environment is a low resource. Actually, here better to catch the output like virtiofsd that will catch the `Waiting for vhost-user socket connection`, but there is no output to catch so I add a sleep 1 second to make sure it boots up.
> 
> Thanks.

Reply:
Ah, I see, my fault. Keep using sleep please. Thanks.